hal/halx86/generic/kdpci.c

```diff
@@ -0,0 +1,532 @@
+    PciDevice->SubClass = PciConfig.SubClass;
+    PciDevice->ProgIf = PciConfig.ProgIf;
+
+    /* Enable decodes */
+    PciConfig.Command |= (PCI_ENABLE_MEMORY_SPACE |
+                          PCI_ENABLE_IO_SPACE |
+                          PCI_ENABLE_BUS_MASTER);
+    HalpPhase0SetPciDataByOffset(PciBus,
+                                 PciSlot,
+                                 &PciConfig.Command,
+                                 FIELD_OFFSET(PCI_COMMON_HEADER, Command),
+                                 sizeof(USHORT));
+
+    return TRUE;
+}
+
+static
+CODE_SEG("INIT")
+BOOLEAN
+HalpMatchDebuggingDevice(
+    _In_ PDEBUG_DEVICE_DESCRIPTOR PciDevice,
+    _In_ ULONG PciBus,
+    _In_ PCI_SLOT_NUMBER PciSlot,
+    _In_ PPCI_COMMON_HEADER PciConfig)
+{
+    /* Check if we weren't given a specific device location */
+    if (PciDevice->Bus == 0xFFFFFFFF &&
+        PciDevice->Slot == 0xFFFFFFFF)
+    {
+        if (PciDevice->DeviceID == 0xFFFF &&
+            PciDevice->VendorID == 0xFFFF)
+        {
+            if (PciDevice->BaseClass == PciConfig->BaseClass &&
+                PciDevice->SubClass == PciConfig->SubClass)
+            {
+                if (PciDevice->ProgIf == 0xFF ||
+                    PciDevice->ProgIf == PciConfig->ProgIf)
+                {
+                    return TRUE;
+                }
+            }
+        }
+        else if (PciDevice->DeviceID == PciConfig->DeviceID &&
+                 PciDevice->VendorID == PciConfig->VendorID)
+        {
+            return TRUE;
+        }
+    }
+    else if (PciDevice->Bus == PciBus &&
+             PciDevice->Slot == PciSlot.u.AsULONG)
+    {
+         return TRUE;
+    }
+
+    return FALSE;
+}
+
+static
+CODE_SEG("INIT")
+BOOLEAN
+HalpFindMatchingDebuggingDevice(
+    _In_ PDEBUG_DEVICE_DESCRIPTOR PciDevice)
+{
+    ULONG BusNumber, DeviceNumber, FunctionNumber;
+
+    for (BusNumber = 0; BusNumber < 0xFF; ++BusNumber)
+    {
+        for (DeviceNumber = 0; DeviceNumber < PCI_MAX_DEVICES; ++DeviceNumber)
+        {
+            for (FunctionNumber = 0; FunctionNumber < PCI_MAX_FUNCTION; ++FunctionNumber)
+            {
+                ULONG Bytes;
+                PCI_SLOT_NUMBER PciSlot;
+                PCI_COMMON_HEADER PciConfig;
+
+                PciSlot.u.bits.DeviceNumber = DeviceNumber;
+                PciSlot.u.bits.FunctionNumber = FunctionNumber;
+                PciSlot.u.bits.Reserved = 0;
+                Bytes = HalpPhase0GetPciDataByOffset(BusNumber,
+                                                     PciSlot,
+                                                     &PciConfig,
+                                                     0,
+                                                     PCI_COMMON_HDR_LENGTH);
+                if (Bytes != PCI_COMMON_HDR_LENGTH ||
+                    PciConfig.VendorID == PCI_INVALID_VENDORID ||
+                    PciConfig.VendorID == 0)
+                {
+                    if (FunctionNumber == 0)
```

Comment:
Can you explain what's happening here?

Reply:
This speeds up the PCI bus scanning
https://git.reactos.org/?p=reactos.git;a=blob;f=drivers/bus/pci/fdo.c;h=2eb5ef2012fa8a5957a1ef3532b97965812e13d0#l97
https://cgit.freebsd.org/src/tree/sys/dev/pci/pci.c#n4185

Reply:
Then you need to add a code comment explaining that.